pallets/randomness/src/types.rs

```diff
@@ -0,0 +1,349 @@
+			),
+		}
+	}
+}
+
+impl<T: Config> From<RequestInfo<T>> for RequestType<T> {
+	fn from(other: RequestInfo<T>) -> RequestType<T> {
+		match other {
+			RequestInfo::BabeEpoch(epoch, _) => RequestType::BabeEpoch(epoch),
+			RequestInfo::Local(block, _) => RequestType::Local(block),
+		}
+	}
+}
+
+#[derive(PartialEq, Clone, Encode, Decode, RuntimeDebug, TypeInfo)]
+/// Raw randomness snapshot, the unique value for a `RequestType` in `RandomnessResults` map
+pub struct RandomnessResult<Hash> {
+	/// Randomness once available
+	pub randomness: Option<Hash>,
+	/// Number of randomness requests for the type
+	pub request_count: u64,
+}
+
+impl<Hash: Clone> RandomnessResult<Hash> {
+	pub fn new() -> RandomnessResult<Hash> {
+		RandomnessResult {
+			randomness: None,
+			request_count: 1u64,
+		}
+	}
+	/// Increment request count
+	pub fn increment_request_count(&self) -> Self {
+		let new_request_count = self.request_count.saturating_add(1u64);
+		RandomnessResult {
+			randomness: self.randomness.clone(),
+			request_count: new_request_count,
+		}
+	}
+	/// Returns whether successfully decremented
+	/// Failure implies the randomness result should be removed from storage
+	pub fn decrement_request_count(&self) -> Option<Self> {
+		if let Some(new_request_count) = self.request_count.checked_sub(1u64) {
+			Some(RandomnessResult {
+				randomness: self.randomness.clone(),
+				request_count: new_request_count,
+			})
+		} else {
+			None
+		}
+	}
+}
+
+#[derive(PartialEq, Clone, Encode, Decode, RuntimeDebug, TypeInfo)]
+/// Input arguments to request randomness
+pub struct Request<Balance, Info> {
+	/// Fee is returned to this account upon execution
+	pub refund_address: H160,
+	/// Contract that consumes the randomness
+	pub contract_address: H160,
+	/// Fee to pay for execution
+	pub fee: Balance,
+	/// Gas limit for subcall
+	pub gas_limit: u64,
+	/// Number of random outputs requested
+	pub num_words: u8,
+	/// Salt to use once randomness is ready
+	pub salt: H256,
+	/// Details regarding request type
+	pub info: Info,
+}
+
+impl<T: Config> From<Request<BalanceOf<T>, RequestType<T>>>
+	for Request<BalanceOf<T>, RequestInfo<T>>
+{
+	fn from(other: Request<BalanceOf<T>, RequestType<T>>) -> Request<BalanceOf<T>, RequestInfo<T>> {
+		Request {
+			refund_address: other.refund_address,
+			contract_address: other.contract_address,
+			fee: other.fee,
+			gas_limit: other.gas_limit,
+			num_words: other.num_words,
+			salt: other.salt,
+			info: other.info.into(),
+		}
+	}
+}
+
+impl<T: Config> Request<BalanceOf<T>, RequestInfo<T>> {
+	pub fn is_expired(&self) -> bool {
+		match self.info {
+			RequestInfo::BabeEpoch(_, expires) => RelayEpoch::<T>::get() >= expires,
+			RequestInfo::Local(_, expires) => frame_system::Pallet::<T>::block_number() >= expires,
+		}
+	}
+	pub fn can_be_fulfilled(&self) -> bool {
+		match self.info {
+			RequestInfo::BabeEpoch(epoch_due, _) => {
+				epoch_due <= T::BabeDataGetter::get_epoch_index()
```

Comment:
Why in is_expired we read it from our storage, and here we read it from the cumulus proof?

Reply:
We should be reading it from storage here, it's a good catch